src/shared/sol-util-linux.c

```diff
@@ -45,3 +59,107 @@ sol_util_timespec_get_realtime(struct timespec *t)
 {
     return clock_gettime(CLOCK_REALTIME, t);
 }
+
+static int
+rand_gen(struct sol_uuid *p)
+{
+    struct sol_buffer buffer = SOL_BUFFER_INIT_EMPTY;
+    int n = sizeof(*p);
+    int fd = -1;
+    int r;
+
+#ifdef HAVE_GETRANDOM
+    r = syscall(SYS_getrandom, p, n, 0);
+    if (r == n)
+        return 0;
+
+    if (r < 0) {
+        if (errno != EAGAIN) { /* fallback to /dev/urandom on EAGAIN
+                                * -- good enough for our uses */
+            return -errno;
+        }
+    } else
+        return -ENODATA;
+#endif
+
+    fd = open("/dev/urandom", O_RDONLY | O_CLOEXEC | O_NOCTTY);
+    if (fd < 0)
+        return errno == ENOENT ? -ENOSYS : -errno;
+
+    r = sol_buffer_ensure(&buffer, n);
+    if (r < 0) {
+        close(fd);
+        return -ENOMEM;
+    }
+    r = sol_util_fill_buffer(fd, &buffer, n);
+    if (r < 0)
+        goto end;
+
+    memcpy(p, buffer.data, n);
+
+end:
+    close(fd);
+    sol_buffer_fini(&buffer);
+    return r;
+}
+
+static struct sol_uuid
+assert_uuid_v4(struct sol_uuid id)
+{
+    id.bytes[6] = (id.bytes[6] & 0x0F) | 0x40;
+    id.bytes[8] = (id.bytes[8] & 0x3F) | 0x80;
+
+    return id;
+}
+
+static int
+uuid_gen(struct sol_uuid *ret)
+{
+    struct sol_uuid t;
+    int r;
+
+    SOL_NULL_CHECK(ret, -EINVAL);
+
+    r = rand_gen(&t);
+    SOL_INT_CHECK(r, < 0, r);
+
+    *ret = assert_uuid_v4(t);
+    return 0;
+}
+
+// 37 = 2 * 16 (chars) + 4 (hyphens) + 1 (\0)
+int
+sol_util_uuid_gen(bool upcase,
+    bool with_hyphens,
+    char id[37])
+{
+    struct sol_str_slice hyphen = SOL_STR_SLICE_LITERAL("-");
+    /* hyphens on positions 8, 13, 18, 23 (from 0) */
+    const int hyphens_pos[] = { 8, 13, 18, 23 };
+    struct sol_uuid uuid;
+    unsigned i;
+    int r;
+
+    struct sol_buffer buf = SOL_BUFFER_INIT_EMPTY;
```

Comment:
Can't this buffer be initialized passing `id` as the data pointer, to avoid the memcpy() below?